### ISSUE TITLE
Rancher agent with only worker role fails to rejoin cluster configured with aws cloud provider after it is rebooted

### ISSUE THREAD

**Issue by garyellis** [rancher/rancher#17431]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Steps to reproduce (least amount of steps as possible):**
1. Create single or HA rancher control plane in AWS in any region where ROS AMI is present.
2. Create a node template for ROS 1.5.0 (or whatever current version is)
3. Add cluster -> from nodes in an infrastructure provider -> Amazon EC2 -> create cluster using ROS node template and with at least one system with worker role only. Configure the Kubernetes cloud provider as Amazon. Wait for cluster provisioning to complete.
4. ssh into worker and reboot.
5. Expecting the rancher worker node agent launched at system startup to repopulate kube certs and kubeconfigs;  kube-proxy and kubelet should start automatically without user interaction.

**Result:**
* Rancher UI reports node unavailable with message `Kubelet stopped posting node status`.
* kubelet and kube-proxy are in a restart loop.
```
F0114 18:53:51.283237    5658 server.go:245] unable to load client CA file /etc/kubernetes/ssl/kube-ca.pem: open /etc/kubernetes/ssl/kube-ca.pem: no such file or directory
```
* Rancher-agent container (launched outside of Kubernetes) fails to connect to rancher `/v3/connect` endpoint with error message `websocket: bad handshake`
```
time="2019-01-14T18:49:52Z" level=info msg="Connecting to proxy" url="wss://gellis-foo.bar/v3/connect"
time="2019-01-14T18:49:52Z" level=error msg="Failed to connect to proxy" error="websocket: bad handshake"
time="2019-01-14T18:49:52Z" level=error msg="Failed to connect to proxy" error="websocket: bad handshake"
```

**Other details that may be helpful:**
On ROS `/etc/kubernetes` is ephemeral. The worker node depends on the rancher agent started at boot to re-populate kubeconfig and kube certificates in order to startup the kubelet. This is not happening with ROS nodes in AWS. Specifically, the docker run command is setting the CATTLE_NODE_NAME to short hostname on worker node reboot. This is causing "bad handshake" to rancher server's websocket tunnel server. Connection to rancher server will succeed when using the node id as set in CRD or the AWS fqdn.

**Workarounds:**
Start the rancher agent using a valid  `CATTLE_NODE_NAME/-n` value to establish a websocket connection to the rancher server. The rancher nodesyncer controller will run shortly after connection is created and re-populate kubelet, kube-proxy client certificates, kube-ca and kubeconfig files. The kubelet and kube-proxy containers will then startup as expected. The kubelet will start the cattle-agent daemonset and any other workloads scheduled to run on the node.

To start the rancher agent manually follow the steps defined in the following workarounds:

**Workaorund 1 - Rerun the docker run command originally run by rancher provisioning process:**
  * Check if `/var/log/secure` is present and copy the original command executed by the rancher server node syncer process. 
  * Run the command  by hand and confirm the log message `Starting plan monitor` is written to stdout.

  * The `CATTLE_NODE_NAME` value `m-xxxxx` will be recognized by rancher server and websocket connection will be established.
```
rancher_agent=$(sed 's/.*COMMAND=//;/rancher-agent/t;d' /var/log/secure*|head -1)
docker logs -f $($rancher_agent)
INFO: Arguments: --server https://gellis-foo.bar --token REDACTED --ca-checksum 0a97bd99fc206f5ec00e5df0a7cbd5c7a34bf210cf5302d36cc397dfe215fed9 -r -n m-xgmcg
INFO: Environment: CATTLE_ADDRESS=10.248.128.37 CATTLE_AGENT_CONNECT=true CATTLE_INTERNAL_ADDRESS= CATTLE_NODE_NAME=m-xgmcg CATTLE_SERVER=https://gellis-foo.bar CATTLE_TOKEN=REDACTED
INFO: Using resolv.conf: domain us-west-2.compute.internal nameserver 10.248.128.2
INFO: https://gellis-foo/ping is accessible
INFO: Value from https://gellis-foo.bar/v3/settings/cacerts is an x509 certificate
time="2019-01-14T19:06:04Z" level=info msg="Rancher agent version v2.1.5 is starting"
time="2019-01-14T19:06:04Z" level=info msg="Option customConfig=map[address:10.248.128.37 internalAddress: roles:[] label:map[]]"
time="2019-01-14T19:06:04Z" level=info msg="Option etcd=false"
time="2019-01-14T19:06:04Z" level=info msg="Option controlPlane=false"
time="2019-01-14T19:06:04Z" level=info msg="Option worker=false"
time="2019-01-14T19:06:04Z" level=info msg="Option requestedHostname=m-xgmcg"
time="2019-01-14T19:06:04Z" level=info msg="Listening on /tmp/log.sock"
time="2019-01-14T19:06:04Z" level=info msg="Connecting to wss://gellis-foo.bar/v3/connect with token 9xfmc6ps9r4sxklsbqzn5gt9bbvpmp4jtjdtdjnftbgtnz2ds4zbhx"
time="2019-01-14T19:06:04Z" level=info msg="Connecting to proxy" url="wss://gellis-foo.bar/v3/connect"
time="2019-01-14T19:06:06Z" level=info msg="Starting plan monitor"
```

**Workaround 2 - Populate the rancher agent docker run command by hand from the failing container.**
  * Populate the rancher agent container volume mounts and start args from the current failing container. Use the AWS fqdn as the cattle node parameter. 
  * An example of setting up docker run command is:
```bash
failing_agent=$(docker ps|grep rancher-agent|awk '{print $1}'|tail -1)
args=$(docker inspect $failing_agent | jq -r '.[].Args|join (" ")')
volumes=$(docker inspect $failing_agent | jq -r '.[].HostConfig | "-v " + .Binds[0] + " -v " + .Binds[1]')
fqdn=$(wget -qO- http://169.254.169.254/latest/meta-data/local-hostname/)
cmd=$(echo docker run -d --privileged --net=host $volumes rancher/rancher-agent:v2.1.5 $args -n $fqdn)

echo $cmd
docker logs -f $($cmd)
```

**Environment information**
OS RancherOS 1.5.0
Rancher version: v2.1.5
Installation option: HA

**Cluster information**
Cluster type (Hosted/Infrastructure Provider/Custom/Imported): Infrastructure provider AWS
Machine type (cloud/VM/metal) and specifications (CPU/memory): Cloud

**Comment by akabdog** [comment]
I'm seeing identical issues with nodes in Rancher 2.1.6
I also tested if a software reboot or hardware powercycle made a difference, the issue happens regardless.

**Comment by superseb** [comment]
There are a few thins off here. The agent launched outside of Kubernetes is launched once (well, 3 containers using the agent image) where one is `share-mnt` which relaunches the agent container on reboot/restart with `--write-certs-only` that will write the certificates before starting the components. After a node has been added to the cluster, the agent container launched outside of Kubernetes gets replaced by the one that is launched using the DaemonSet in the cluster. So that agent shouldn't be there after it has been successfully added to the cluster. If this is left behind after adding it or something else is going on, we need that info (`docker ps -a` output and logs) to see why that happens.

I've tried reproducing it but haven't been able to using `v2.1.6` and `v2.3.0-rc1`. If you run into it, please share the output of `docker ps -a` and logs from the containers using the agent image.

**Comment by Surakarus** [comment]
This is the result of `docker ps -a` command on a rebooted node
```
CONTAINER ID        IMAGE                                COMMAND                  CREATED             STATUS                            PORTS               NAMES
7fa7635d69b4        rancher/rancher-agent:v2.2.8         "run.sh --server htt…"   15 minutes ago      Up 15 minutes                                         zealous_einstein
a6e76c40f4c7        a89b45f36d5e                         "start_runit"            4 hours ago         Exited (0) 16 minutes ago                             k8s_calico-node_canal-6xz4d_kube-system_878f6a81-d3cf-11e9-a04c-021ac9ffe692_2
8a72ea7ac77c        2b37f252629b                         "/entrypoint.sh /ngi…"   4 hours ago         Exited (143) 16 minutes ago                           k8s_nginx-ingress-controller_nginx-ingress-controller-cbh5c_ingress-nginx_8d85cba9-d3cf-11e9-a04c-021ac9ffe692_3
8acf91b2c0be        f0fad859c909                         "/opt/bin/flanneld -…"   4 hours ago         Exited (255) 15 minutes ago                           k8s_kube-flannel_canal-6xz4d_kube-system_878f6a81-d3cf-11e9-a04c-021ac9ffe692_2
212195d3d227        2b37f252629b                         "/entrypoint.sh /ngi…"   4 hours ago         Exited (143) 4 hours ago                              k8s_nginx-ingress-controller_nginx-ingress-controller-cbh5c_ingress-nginx_8d85cba9-d3cf-11e9-a04c-021ac9ffe692_2
120357939e55        f0fad859c909                         "/opt/bin/flanneld -…"   4 hours ago         Exited (1) 4 hours ago                                k8s_kube-flannel_canal-6xz4d_kube-system_878f6a81-d3cf-11e9-a04c-021ac9ffe692_1
5b84c133c96f        a89b45f36d5e                         "start_runit"            4 hours ago         Exited (137) 4 hours ago                              k8s_calico-node_canal-6xz4d_kube-system_878f6a81-d3cf-11e9-a04c-021ac9ffe692_1
874eca352df1        d499bdb8fd09                         "run.sh"                 4 hours ago         Exited (2) 16 minutes ago                             k8s_agent_cattle-node-agent-c4skj_cattle-system_878c0d7b-d3cf-11e9-a04c-021ac9ffe692_1
60ee56ec4b5b        d531d047a4e2                         "/install-cni.sh"        4 hours ago         Exited (0) 4 hours ago                                k8s_install-cni_canal-6xz4d_kube-system_878f6a81-d3cf-11e9-a04c-021ac9ffe692_1
01ede296f5eb        rancher/pause:3.1                    "/pause"                 4 hours ago         Exited (0) 16 minutes ago                             k8s_POD_cattle-node-agent-c4skj_cattle-system_878c0d7b-d3cf-11e9-a04c-021ac9ffe692_1
6f79d70cad12        rancher/pause:3.1                    "/pause"                 4 hours ago         Exited (0) 16 minutes ago                             k8s_POD_canal-6xz4d_kube-system_878f6a81-d3cf-11e9-a04c-021ac9ffe692_1
040cbfe81a30        rancher/pause:3.1                    "/pause"                 4 hours ago         Exited (0) 16 minutes ago                             k8s_POD_nginx-ingress-controller-cbh5c_ingress-nginx_8d85cba9-d3cf-11e9-a04c-021ac9ffe692_1
9042f9e4ad55        d499bdb8fd09                         "run.sh"                 26 hours ago        Exited (2) 4 hours ago                                k8s_agent_cattle-node-agent-c4skj_cattle-system_878c0d7b-d3cf-11e9-a04c-021ac9ffe692_0
a226f39adf43        rancher/pause:3.1                    "/pause"                 26 hours ago        Exited (0) 4 hours ago                                k8s_POD_cattle-node-agent-c4skj_cattle-system_878c0d7b-d3cf-11e9-a04c-021ac9ffe692_0
92bfece20d91        rancher/hyperkube:v1.13.5-rancher1   "/opt/rke-tools/entr…"   26 hours ago        Restarting (255) 21 seconds ago                       kubelet
7300a396d000        rancher/hyperkube:v1.13.5-rancher1   "/opt/rke-tools/entr…"   26 hours ago        Restarting (255) 58 seconds ago                       kube-proxy
b3a60da975a0        rancher/rke-tools:v0.1.28            "nginx-proxy CP_HOST…"   26 hours ago        Up 15 minutes                                         nginx-proxy
6f42b04278e9        rancher/rancher-agent:v2.2.8         "run.sh -- share-roo…"   26 hours ago        Exited (0) 15 minutes ago                             share-mnt
b6c4fbc37588        rancher/rke-tools:v0.1.28            "/bin/bash"              26 hours ago        Exited (0) 16 minutes ago                             service-sidekick
```
And this is the logs of rancher-agent container:
```
INFO: Arguments: --server https://rancher.example.com --token REDACTED --no-register --only-write-certs
INFO: Environment: CATTLE_ADDRESS=10.0.4.168 CATTLE_AGENT_CONNECT=true CATTLE_INTERNAL_ADDRESS= CATTLE_NODE_NAME=ip-10-0-4-168 CATTLE_SERVER=https://rancher.example.com CATTLE_TOKEN=REDACTED CATTLE_WRITE_CERT_ONLY=true
INFO: Using resolv.conf: domain us-east-2.compute.internal nameserver 10.0.0.2
INFO: https://rancher.example.com/ping is accessible
INFO: rancher.example.com resolves to 172.31.4.100
time="2019-09-11T15:11:17Z" level=info msg="Rancher agent version v2.2.8 is starting"
time="2019-09-11T15:11:17Z" level=info msg="Option customConfig=map[address:10.0.4.168 internalAddress: label:map[] roles:[]]"
time="2019-09-11T15:11:17Z" level=info msg="Option etcd=false"
time="2019-09-11T15:11:17Z" level=info msg="Option controlPlane=false"
time="2019-09-11T15:11:17Z" level=info msg="Option worker=false"
time="2019-09-11T15:11:17Z" level=info msg="Option requestedHostname=ip-10-0-4-168"
time="2019-09-11T15:11:17Z" level=info msg="Listening on /tmp/log.sock"
time="2019-09-11T15:11:17Z" level=info msg="Connecting to wss://rancher.example.com/v3/connect with token REDACTED"
time="2019-09-11T15:11:17Z" level=info msg="Connecting to proxy" url="wss://rancher.example.com/v3/connect"
time="2019-09-11T15:11:18Z" level=error msg="Failed to connect to proxy" error="websocket: bad handshake"
time="2019-09-11T15:11:18Z" level=error msg="Failed to connect to proxy" error="websocket: bad handshake"
```

**Comment by superseb** [comment]
Can you also share output of `docker info` and the output of `View in API` in Rancher from the cluster this node is a part of (remove any sensitive info before posting)?

**Comment by Surakarus** [comment]
Output of `docker info`:
```
Containers: 19
 Running: 3
 Paused: 0
 Stopped: 16
Images: 8
Server Version: 18.09.8
Storage Driver: overlay
 Backing Filesystem: extfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 894b81a4b802e4eb2a91d1ce216b8817763c29fb
runc version: 425e105d5a03fabd737a126ad93d62a9eeede87f
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.14.85-rancher
Operating System: RancherOS v1.5.1
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.67GiB
Name: ip-10-0-4-168.us-east-2.compute.internal
ID: MG64:CVO6:3545:KH42:VECC:2GGA:ROIL:7FAI:2FNA:V7QV:EOAK:3TYI
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
 provider=amazonec2
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine

WARNING: the overlay storage-driver is deprecated, and will be removed in a future release.
```
<details>
<summary>
Output of "View in API":
</summary>
```
{

    "actions": {
        "backupEtcd": "…/v3/clusters/c-dc8c4?action=backupEtcd",
        "enableMonitoring": "…/v3/clusters/c-dc8c4?action=enableMonitoring",
        "exportYaml": "…/v3/clusters/c-dc8c4?action=exportYaml",
        "generateKubeconfig": "…/v3/clusters/c-dc8c4?action=generateKubeconfig",
        "importYaml": "…/v3/clusters/c-dc8c4?action=importYaml",
        "restoreFromEtcdBackup": "…/v3/clusters/c-dc8c4?action=restoreFromEtcdBackup",
        "rotateCertificates": "…/v3/clusters/c-dc8c4?action=rotateCertificates"
    },
    "agentImage": "rancher/rancher-agent:v2.2.8",
    "allocatable": {
        "cpu": "8",
        "memory": "32648076Ki",
        "pods": "220"
    },
    "annotations": {
        "authz.management.cattle.io/creator-role-bindings": "{\"created\":[\"cluster-owner\"],\"required\":[\"cluster-owner\"]}",
        "lifecycle.cattle.io/create.cluster-agent-controller-cleanup": "true",
        "lifecycle.cattle.io/create.cluster-provisioner-controller": "true",
        "lifecycle.cattle.io/create.cluster-scoped-gc": "true",
        "lifecycle.cattle.io/create.mgmt-cluster-rbac-remove": "true",
        "provisioner.cattle.io/ke-driver-update": "updated"
    },
    "apiEndpoint": "https://10.0.4.240:6443",
    "appliedEnableNetworkPolicy": false,
    "appliedPodSecurityPolicyTemplateId": "",
    "appliedSpec": {
        "defaultClusterRoleForProjectMembers": null,
        "defaultPodSecurityPolicyTemplateId": null,
        "description": "",
        "desiredAgentImage": "",
        "desiredAuthImage": "",
        "displayName": "test1",
        "dockerRootDir": "/var/lib/docker",
        "enableClusterAlerting": false,
        "enableClusterMonitoring": false,
        "enableNetworkPolicy": false,
        "internal": false,
        "localClusterAuthEndpoint": {
            "enabled": true,
            "type": "/v3/schemas/localClusterAuthEndpoint"
        },
        "rancherKubernetesEngineConfig": {
            "addonJobTimeout": 30,
            "authentication": {
                "strategy": "x509|webhook",
                "type": "/v3/schemas/authnConfig"
            },
            "authorization": {
                "type": "/v3/schemas/authzConfig"
            },
            "bastionHost": {
                "sshAgentAuth": false,
                "type": "/v3/schemas/bastionHost"
            },
            "cloudProvider": {
                "awsCloudProvider": {
                    "global": {
                        "disable-security-group-ingress": false,
                        "disable-strict-zone-check": false,
                        "elb-security-group": "",
                        "kubernetes-cluster-id": "",
                        "kubernetes-cluster-tag": "",
                        "role-arn": "",
                        "routetable-id": "",
                        "subnet-id": "",
                        "type": "/v3/schemas/globalAwsOpts",
                        "vpc": "",
                        "zone": ""
                    },
                    "type": "/v3/schemas/awsCloudProvider"
                },
                "name": "aws",
                "type": "/v3/schemas/cloudProvider"
            },
            "ignoreDockerVersion": true,
            "ingress": {
                "provider": "nginx",
                "type": "/v3/schemas/ingressConfig"
            },
            "kubernetesVersion": "v1.13.5-rancher1-3",
            "monitoring": {
                "provider": "metrics-server",
                "type": "/v3/schemas/monitoringConfig"
            },
            "network": {
                "options": {
                    "flannel_backend_type": "vxlan"
                },
                "plugin": "canal",
                "type": "/v3/schemas/networkConfig"
            },
            "nodes": [ 3 items
                {
                    "address": "10.0.4.139",
                    "hostnameOverride": "k8s-test1-worker1",
                    "internalAddress": "10.0.4.139",
                    "labels": {
                        "cattle.io/creator": "norman"
                    },
                    "nodeId": "c-dc8c4:m-5v4nr",
                    "port": "22",
                    "role": [
                        "worker"
                    ],
                    "sshAgentAuth": false,
                    "type": "/v3/schemas/rkeConfigNode",
                    "user": "rancher"
                },
                {
                    "address": "10.0.4.240",
                    "hostnameOverride": "k8s-test1-master1",
                    "internalAddress": "10.0.4.240",
                    "labels": {
                        "cattle.io/creator": "norman"
                    },
                    "nodeId": "c-dc8c4:m-8xj9m",
                    "port": "22",
                    "role": [ 2 items
                        "controlplane",
                        "etcd"
                    ],
                    "sshAgentAuth": false,
                    "type": "/v3/schemas/rkeConfigNode",
                    "user": "rancher"
                },
                {
                    "address": "10.0.4.168",
                    "hostnameOverride": "k8s-test1-worker2",
                    "internalAddress": "10.0.4.168",
                    "labels": {
                        "cattle.io/creator": "norman"
                    },
                    "nodeId": "c-dc8c4:m-m64zm",
                    "port": "22",
                    "role": [
                        "worker"
                    ],
                    "sshAgentAuth": false,
                    "type": "/v3/schemas/rkeConfigNode",
                    "user": "rancher"
                }
            ],
            "restore": {
                "restore": false,
                "type": "/v3/schemas/restoreConfig"
            },
            "services": {
                "etcd": {
                    "backupConfig": {
                        "enabled": true,
                        "intervalHours": 12,
                        "retention": 6,
                        "s3BackupConfig": null,
                        "type": "/v3/schemas/backupConfig"
                    },
                    "creation": "12h",
                    "extraArgs": {
                        "election-timeout": "5000",
                        "heartbeat-interval": "500"
                    },
                    "retention": "72h",
                    "snapshot": false,
                    "type": "/v3/schemas/etcdService"
                },
                "kubeApi": {
                    "alwaysPullImages": false,
                    "podSecurityPolicy": false,
                    "serviceNodePortRange": "30000-32767",
                    "type": "/v3/schemas/kubeAPIService"
                },
                "kubeController": {
                    "type": "/v3/schemas/kubeControllerService"
                },
                "kubelet": {
                    "failSwapOn": false,
                    "type": "/v3/schemas/kubeletService"
                },
                "kubeproxy": {
                    "type": "/v3/schemas/kubeproxyService"
                },
                "scheduler": {
                    "type": "/v3/schemas/schedulerService"
                },
                "type": "/v3/schemas/rkeConfigServices"
            },
            "sshAgentAuth": false,
            "type": "/v3/schemas/rancherKubernetesEngineConfig"
        },
        "type": "/v3/schemas/clusterSpec"
    },
    "authImage": "rancher/kube-api-auth:v0.1.3",
    "baseType": "cluster",
    "caCert": "CERTIFICATE",
    "capabilities": {
        "ingressCapabilities": [
            {
                "customDefaultBackend": false,
                "ingressProvider": "Nginx",
                "type": "/v3/schemas/ingressCapabilities"
            }
        ],
        "loadBalancerCapabilities": {
            "enabled": true,
            "healthCheckSupported": true,
            "protocolsSupported": [
                "TCP"
            ],
            "provider": "ELB",
            "type": "/v3/schemas/loadBalancerCapabilities"
        },
        "nodePoolScalingSupported": true,
        "nodePortRange": "30000-32767",
        "type": "/v3/schemas/capabilities"
    },
    "capacity": {
        "cpu": "8",
        "memory": "32852876Ki",
        "pods": "220"
    },
    "certificatesExpiration": {
        "kube-apiserver": {
            "expirationDate": "2029-05-26T13:25:19Z",
            "type": "/v3/schemas/certExpiration"
        },
        "kube-ca": {
            "expirationDate": "2029-05-26T13:25:18Z",
            "type": "/v3/schemas/certExpiration"
        },
        "kube-controller-manager": {
            "expirationDate": "2029-05-26T13:25:19Z",
            "type": "/v3/schemas/certExpiration"
        },
        "kube-etcd-10-0-4-240": {
            "expirationDate": "2029-05-26T13:25:19Z",
            "type": "/v3/schemas/certExpiration"
        },
        "kube-node": {
            "expirationDate": "2029-05-26T13:25:20Z",
            "type": "/v3/schemas/certExpiration"
        },
        "kube-proxy": {
            "expirationDate": "2029-05-26T13:25:19Z",
            "type": "/v3/schemas/certExpiration"
        },
        "kube-scheduler": {
            "expirationDate": "2029-05-26T13:25:20Z",
            "type": "/v3/schemas/certExpiration"
        }
    },
    "componentStatuses": [ 3 items
        {
            "conditions": [
                {
                    "message": "ok",
                    "status": "True",
                    "type": "Healthy"
                }
            ],
            "name": "controller-manager",
            "type": "/v3/schemas/clusterComponentStatus"
        },
        {
            "conditions": [
                {
                    "message": "{\"health\": \"true\"}",
                    "status": "True",
                    "type": "Healthy"
                }
            ],
            "name": "etcd-0",
            "type": "/v3/schemas/clusterComponentStatus"
        },
        {
            "conditions": [
                {
                    "message": "ok",
                    "status": "True",
                    "type": "Healthy"
                }
            ],
            "name": "scheduler",
            "type": "/v3/schemas/clusterComponentStatus"
        }
    ],
    "conditions": [ 18 items
        {
            "status": "True",
            "type": "Pending"
        },
        {
            "lastUpdateTime": "2019-05-29T13:27:22Z",
            "status": "True",
            "type": "Provisioned"
        },
        {
            "lastUpdateTime": "2019-05-29T13:27:37Z",
            "status": "True",
            "type": "Waiting"
        },
        {
            "lastUpdateTime": "2019-05-29T13:22:38Z",
            "status": "True",
            "type": "BackingNamespaceCreated"
        },
        {
            "lastUpdateTime": "2019-05-29T13:22:38Z",
            "status": "True",
            "type": "DefaultProjectCreated"
        },
        {
            "lastUpdateTime": "2019-05-29T13:22:38Z",
            "status": "True",
            "type": "SystemProjectCreated"
        },
        {
            "lastUpdateTime": "2019-05-29T13:22:38Z",
            "status": "True",
            "type": "InitialRolesPopulated"
        },
        {
            "lastUpdateTime": "2019-05-29T13:22:38Z",
            "status": "True",
            "type": "CreatorMadeOwner"
        },
        {
            "lastUpdateTime": "2019-05-29T13:22:38Z",
            "status": "True",
            "type": "NoDiskPressure"
        },
        {
            "lastUpdateTime": "2019-05-29T13:22:38Z",
            "status": "True",
            "type": "NoMemoryPressure"
        },
        {
            "lastUpdateTime": "2019-05-29T13:27:23Z",
            "status": "True",
            "type": "GlobalAdminsSynced"
        },
        {
            "lastUpdateTime": "2019-05-29T13:27:23Z",
            "status": "False",
            "type": "AlertingEnabled"
        },
        {
            "lastUpdateTime": "2019-05-29T13:27:27Z",
            "status": "True",
            "type": "SystemAccountCreated"
        },
        {
            "lastUpdateTime": "2019-05-29T13:27:29Z",
            "status": "True",
            "type": "AgentDeployed"
        },
        {
            "lastUpdateTime": "2019-05-29T13:27:29Z",
            "status": "False",
            "type": "PrometheusOperatorDeployed"
        },
        {
            "lastUpdateTime": "2019-09-03T11:43:05Z",
            "status": "True",
            "type": "Ready"
        },
        {
            "lastUpdateTime": "2019-05-29T13:28:27Z",
            "status": "True",
            "type": "Updated"
        },
        {
            "lastUpdateTime": "2019-05-29T13:27:48Z",
            "status": "True",
            "type": "ServiceAccountMigrated"
        }
    ],
    "created": "2019-05-29T13:22:38Z",
    "createdTS": 1559136158000,
    "creatorId": "user-gc6cq",
    "defaultClusterRoleForProjectMembers": null,
    "defaultPodSecurityPolicyTemplateId": null,
    "description": "",
    "desiredAgentImage": "",
    "desiredAuthImage": "",
    "dockerRootDir": "/var/lib/docker",
    "driver": "rancherKubernetesEngine",
    "enableClusterAlerting": false,
    "enableClusterMonitoring": false,
    "enableNetworkPolicy": false,
    "id": "c-dc8c4",
    "internal": false,
    "labels": {
        "cattle.io/creator": "norman"
    },
    "limits": {
        "cpu": "16010m",
        "memory": "18622Mi",
        "pods": "0"
    },
    "links": {
        "clusterAlertGroups": "…/v3/clusters/c-dc8c4/clusteralertgroups",
        "clusterAlertRules": "…/v3/clusters/c-dc8c4/clusteralertrules",
        "clusterAlerts": "…/v3/clusters/c-dc8c4/clusteralerts",
        "clusterCatalogs": "…/v3/clusters/c-dc8c4/clustercatalogs",
        "clusterLoggings": "…/v3/clusters/c-dc8c4/clusterloggings",
        "clusterMonitorGraphs": "…/v3/clusters/c-dc8c4/clustermonitorgraphs",
        "clusterRegistrationTokens": "…/v3/clusters/c-dc8c4/clusterregistrationtokens",
        "clusterRoleTemplateBindings": "…/v3/clusters/c-dc8c4/clusterroletemplatebindings",
        "etcdBackups": "…/v3/clusters/c-dc8c4/etcdbackups",
        "namespaces": "…/v3/clusters/c-dc8c4/namespaces",
        "nodePools": "…/v3/clusters/c-dc8c4/nodepools",
        "nodes": "…/v3/clusters/c-dc8c4/nodes",
        "notifiers": "…/v3/clusters/c-dc8c4/notifiers",
        "persistentVolumes": "…/v3/clusters/c-dc8c4/persistentvolumes",
        "projects": "…/v3/clusters/c-dc8c4/projects",
        "remove": "…/v3/clusters/c-dc8c4",
        "self": "…/v3/clusters/c-dc8c4",
        "shell": "wss://rancher.example.com/v3/clusters/c-dc8c4?shell=true",
        "storageClasses": "…/v3/clusters/c-dc8c4/storageclasses",
        "subscribe": "…/v3/clusters/c-dc8c4/subscribe",
        "templates": "…/v3/clusters/c-dc8c4/templates",
        "tokens": "…/v3/clusters/c-dc8c4/tokens",
        "update": "…/v3/clusters/c-dc8c4"
    },
    "localClusterAuthEndpoint": {
        "enabled": true,
        "type": "/v3/schemas/localClusterAuthEndpoint"
    },
    "name": "test1",
    "rancherKubernetesEngineConfig": {
        "addonJobTimeout": 30,
        "authentication": {
            "strategy": "x509|webhook",
            "type": "/v3/schemas/authnConfig"
        },
        "authorization": {
            "type": "/v3/schemas/authzConfig"
        },
        "bastionHost": {
            "sshAgentAuth": false,
            "type": "/v3/schemas/bastionHost"
        },
        "cloudProvider": {
            "awsCloudProvider": {
                "global": {
                    "disable-security-group-ingress": false,
                    "disable-strict-zone-check": false,
                    "elb-security-group": "",
                    "kubernetes-cluster-id": "",
                    "kubernetes-cluster-tag": "",
                    "role-arn": "",
                    "routetable-id": "",
                    "subnet-id": "",
                    "type": "/v3/schemas/globalAwsOpts",
                    "vpc": "",
                    "zone": ""
                },
                "type": "/v3/schemas/awsCloudProvider"
            },
            "name": "aws",
            "type": "/v3/schemas/cloudProvider"
        },
        "ignoreDockerVersion": true,
        "ingress": {
            "provider": "nginx",
            "type": "/v3/schemas/ingressConfig"
        },
        "kubernetesVersion": "v1.13.5-rancher1-3",
        "monitoring": {
            "provider": "metrics-server",
            "type": "/v3/schemas/monitoringConfig"
        },
        "network": {
            "options": {
                "flannel_backend_type": "vxlan"
            },
            "plugin": "canal",
            "type": "/v3/schemas/networkConfig"
        },
        "restore": {
            "restore": false,
            "type": "/v3/schemas/restoreConfig"
        },
        "services": {
            "etcd": {
                "backupConfig": {
                    "enabled": true,
                    "intervalHours": 12,
                    "retention": 6,
                    "s3BackupConfig": null,
                    "type": "/v3/schemas/backupConfig"
                },
                "creation": "12h",
                "extraArgs": {
                    "election-timeout": "5000",
                    "heartbeat-interval": "500"
                },
                "retention": "72h",
                "snapshot": false,
                "type": "/v3/schemas/etcdService"
            },
            "kubeApi": {
                "alwaysPullImages": false,
                "podSecurityPolicy": false,
                "serviceNodePortRange": "30000-32767",
                "type": "/v3/schemas/kubeAPIService"
            },
            "kubeController": {
                "type": "/v3/schemas/kubeControllerService"
            },
            "kubelet": {
                "failSwapOn": false,
                "type": "/v3/schemas/kubeletService"
            },
            "kubeproxy": {
                "type": "/v3/schemas/kubeproxyService"
            },
            "scheduler": {
                "type": "/v3/schemas/schedulerService"
            },
            "type": "/v3/schemas/rkeConfigServices"
        },
        "sshAgentAuth": false,
        "type": "/v3/schemas/rancherKubernetesEngineConfig"
    },
    "requested": {
        "cpu": "1831m",
        "memory": "8620Mi",
        "pods": "40"
    },
    "state": "active",
    "transitioning": "no",
    "transitioningMessage": "",
    "type": "cluster",
    "uuid": "d4775e85-8214-11e9-b620-0242ac160002",
    "version": {
        "buildDate": "2019-03-25T15:19:22Z",
        "compiler": "gc",
        "gitCommit": "2166946f41b36dea2c4626f90a77706f426cdea2",
        "gitTreeState": "clean",
        "gitVersion": "v1.13.5",
        "goVersion": "go1.11.5",
        "major": "1",
        "minor": "13",
        "platform": "linux/amd64",
        "type": "/v3/schemas/info"
    }

}
```
</details>

**Comment by superseb** [comment]
Ok thanks, the missing link here was a configured cloud provider in Kubernetes set to AWS, so it changes the nodename which causes the mismatch.

**Comment by superseb** [comment]
Merging into https://github.com/rancher/rancher/issues/22849

**Comment by bycEEE** [comment]
Also running into this problem. Our setup is created using `terraform` to create AWS infrastructure and RKE cluster is provisioned with the `rke` CLI tool. Our workload clusters are launched with node templates/pools.

Domains have been changed in my logs.
Upon reboot, rancher-agent will fail with the following:
```
time="2019-09-12T18:13:24Z" level=info msg="Connecting to proxy" url="wss://rke.domain.com/v3/connect"
time="2019-09-12T18:13:24Z" level=error msg="Failed to connect to proxy" error="websocket: bad handshake"
time="2019-09-12T18:13:24Z" level=error msg="Failed to connect to proxy" error="websocket: bad handshake"
```

Tested with:
- Kubernetes: `v1.14.6-rancher1-1`
- Rancher agent: Both `v2.2.8` and `v2.3.0-rc3`
- RancherOS: `1.5.4`

Unhealthy `rancher-agent` on worker node:
```
[root@worker-test-dev-a-0 ~]# docker logs c502dacc87bb
INFO: Arguments: --server https://rke.domain.com --token REDACTED --ca-checksum REDACTED --no-register --only-write-certs
INFO: Environment: CATTLE_ADDRESS=10.10.66.7 CATTLE_AGENT_CONNECT=true CATTLE_INTERNAL_ADDRESS= CATTLE_NODE_NAME=worker-test-dev-a-0 CATTLE_SERVER=https://rke.domain.com CATTLE_TOKEN=REDACTED CATTLE_WRITE_CERT_ONLY=true
```

Since `rancher-agent` has exited on a working worker node, logs from a running container:
```
[rancher@worker-test-dev-c-0 ~]$ docker logs cc4d48a3fdb7
INFO: Environment: CATTLE_ADDRESS=10.10.66.80 CATTLE_AGENT_CONNECT=true CATTLE_CA_CHECKSUM=REDACTED CATTLE_CLUSTER=false CATTLE_INTERNAL_ADDRESS= CATTLE_K8S_MANAGED=true CATTLE_NODE_NAME=ip-10-10-66-80.ec2.internal CATTLE_SERVER=https://rke.domain.com
```

Running the same command on the container with `CATTLE_NODE_NAME=worker-test-dev-a-0` changed to `ip-10-10-66-7.ec2.internal` resolves my issue and allows the node to reconnect.

My current workaround is in terraform to template out the userdata:
```
#cloud-config
write_files:
  - path: /etc/rc.local
    permissions: "0755"
    owner: root
    content: |
      #!/bin/bash
      wait-for-docker
      docker pull rancher/curl
      TEMP_NODE_NAME=$(docker run rancher/curl http://169.254.169.254/latest/meta-data/local-ipv4)
      ${rancher_command} -r -n ip-$${TEMP_NODE_NAME//./-}.ec2.internal
```

```
data "template_file" "userdata_worker" {
  template = file("${path.module}/templates/userdata_worker.tmpl")
  vars = {
    rancher_command = rancher2_cluster.cluster.cluster_registration_token[0].node_command
  }
}
```

**Comment by bycEEE** [comment]
@superseb more logs:

```



[root@worker-redacted-f-0 ~]# docker ps -a
CONTAINER ID        IMAGE                                COMMAND                  CREATED             STATUS                            PORTS               NAMES
93f8baccfaef        rancher/rancher-agent:v2.2.8         "run.sh --server htt…"   2 days ago          Up 2 days                                             wonderful_tu
bf28c6970f48        rancher/coreos-flannel               "/opt/bin/flanneld -…"   2 days ago          Exited (255) 2 days ago                               k8s_kube-flannel_canal-66w7j_kube-system_a32c77ab-d672-11e9-bd4c-0a1c983f91b0_0
184e8a01646f        rancher/flannel-cni                  "/install-cni.sh"        2 days ago          Exited (255) 2 days ago                               k8s_install-cni_kube-flannel-fz54w_kube-system_a331fef6-d672-11e9-bd4c-0a1c983f91b0_0
7a32c29703f1        rancher/calico-node                  "start_runit"            2 days ago          Exited (0) 2 days ago                                 k8s_calico-node_canal-66w7j_kube-system_a32c77ab-d672-11e9-bd4c-0a1c983f91b0_0
e3beeb80cfcd        rancher/coreos-flannel               "/opt/bin/flanneld -…"   2 days ago          Exited (255) 2 days ago                               k8s_kube-flannel_kube-flannel-fz54w_kube-system_a331fef6-d672-11e9-bd4c-0a1c983f91b0_0
960b7559f191        rancher/calico-cni                   "/install-cni.sh"        2 days ago          Exited (0) 2 days ago                                 k8s_install-cni_canal-66w7j_kube-system_a32c77ab-d672-11e9-bd4c-0a1c983f91b0_0
9b41eb784222        d499bdb8fd09                         "run.sh"                 2 days ago          Exited (2) 2 days ago                                 k8s_agent_cattle-node-agent-hmdv5_cattle-system_a32c8731-d672-11e9-bd4c-0a1c983f91b0_0
6be1e30746a7        rancher/pause:3.1                    "/pause"                 2 days ago          Exited (0) 2 days ago                                 k8s_POD_cattle-node-agent-hmdv5_cattle-system_a32c8731-d672-11e9-bd4c-0a1c983f91b0_0
5d7ad46053ef        rancher/pause:3.1                    "/pause"                 2 days ago          Exited (0) 2 days ago                                 k8s_POD_kube-flannel-fz54w_kube-system_a331fef6-d672-11e9-bd4c-0a1c983f91b0_0
7964c569e6e8        rancher/pause:3.1                    "/pause"                 2 days ago          Exited (0) 2 days ago                                 k8s_POD_canal-66w7j_kube-system_a32c77ab-d672-11e9-bd4c-0a1c983f91b0_0
6cc8a91c9217        rancher/hyperkube:v1.14.6-rancher1   "/opt/rke-tools/entr…"   2 days ago          Restarting (255) 50 seconds ago                       kube-proxy
8074225c262e        rancher/rke-tools:v0.1.42            "nginx-proxy CP_HOST…"   2 days ago          Up 2 days                                             nginx-proxy
7b36e56073c3        rancher/hyperkube:v1.14.6-rancher1   "/opt/rke-tools/entr…"   2 days ago          Restarting (255) 14 seconds ago                       kubelet
0510d59d277b        rancher/rancher-agent:v2.2.8         "run.sh -- share-roo…"   2 days ago          Exited (0) 2 days ago                                 share-mnt
79bbf6771572        rancher/rke-tools:v0.1.42            "/bin/bash"              2 days ago          Exited (0) 2 days ago                                 service-sidekick
```

```
[root@worker-redacted-f-0 ~]# docker logs 93f8baccfaef
INFO: Arguments: --server https://rancher.redacted.io --token REDACTED --ca-checksum REDACTED --no-register --only-write-certs
INFO: Environment: CATTLE_ADDRESS=10.20.92.121 CATTLE_AGENT_CONNECT=true CATTLE_INTERNAL_ADDRESS= CATTLE_NODE_NAME=worker-redacted-f-0 CATTLE_SERVER=https://rancher.redacted.io CATTLE_TOKEN=REDACTED CATTLE_WRITE_CERT_ONLY=true
INFO: Using resolv.conf: domain ec2.internal nameserver 10.20.0.2
INFO: https://rancher.redacted.io/ping is accessible
INFO: rancher.redacted.io resolves to 10.30.74.233 10.30.72.196 10.30.73.106
INFO: Value from https://rancher.redacted.io/v3/settings/cacerts is an x509 certificate
time="2019-09-13T22:11:19Z" level=info msg="Rancher agent version v2.2.8 is starting"
time="2019-09-13T22:11:19Z" level=info msg="Option requestedHostname=worker-redacted-0"
time="2019-09-13T22:11:19Z" level=info msg="Option customConfig=map[address:10.20.92.121 internalAddress: label:map[] roles:[]]"
time="2019-09-13T22:11:19Z" level=info msg="Option etcd=false"
time="2019-09-13T22:11:19Z" level=info msg="Option controlPlane=false"
time="2019-09-13T22:11:19Z" level=info msg="Option worker=false"
time="2019-09-13T22:11:19Z" level=info msg="Listening on /tmp/log.sock"
time="2019-09-13T22:11:19Z" level=info msg="Connecting to wss://rancher.redacted.io/v3/connect with token <REDACTED>"
time="2019-09-13T22:11:19Z" level=info msg="Connecting to proxy" url="wss://rancher.redacted.io/v3/connect"
time="2019-09-13T22:11:19Z" level=error msg="Failed to connect to proxy" error="websocket: bad handshake"
time="2019-09-13T22:11:19Z" level=error msg="Failed to connect to proxy" error="websocket: bad handshake"
time="2019-09-13T22:11:30Z" level=info msg="Connecting to wss://rancher.redacted.io/v3/connect with token <REDACTED>"
time="2019-09-13T22:11:30Z" level=info msg="Connecting to proxy" url="wss://rancher.redacted.io/v3/connect"
time="2019-09-13T22:11:30Z" level=error msg="Failed to connect to proxy" error="websocket: bad handshake"
time="2019-09-13T22:11:30Z" level=error msg="Failed to connect to proxy" error="websocket: bad handshake"
time="2019-09-13T22:11:40Z" level=info msg="Connecting to wss://rancher.redacted.io/v3/connect with token <REDACTED>"
```

**Comment by superseb** [comment]
This is reproduced and merged into another open issue that will be fixed.
